### PR TITLE
[full-ci] Display resources in `GenericSpace` as tiles

### LINKED ISSUE #6126: Tiles view for file list
- **needs split** will be split into multiple stories fro a better digestion in the sprint:
  1. Tiles-view switch + Tiles
  2.  Slider
  3. Sorting

- **not a gallery-app:** Note the difference between a "gallery-app" and the "tiles view": The tiles view shows the plain file listing in a grid. It shows all files, regardless of their mime type, and does not add specific functionality for certain file types.

# Description

## User Stories

* > As a user I want to switch to a Tiles view so that I get an overview of all files of a folder without having to rely on filenames.

## Value
- improve files navigation & organisation
- "Tiles" will be reused for the visual representation of **"recent files"** and **"Spaces overview"**

## Acceptance Criteria

- **tiles-view switch** Add a switch to switch to a tilesview (from list view)
  - persistence: the current selection of the switch [condensed] [Normal] [Tiles] is persisted through different view (meaning: the viewmode only changes, if the user selects another viewmode)
  - the 3 viewmodes are in the first step only available in (to keep it simple)
    - "Personal"
    - Every single Space
- **tiles-view switch:** if tiles-switch is active, a slider appears
  - **mobile:** slider is not available on smartphone-screens because of missing real estate
  - **mobile:** only 1 button (not a group of 2-3 buttons), that toggles between standard-list and tiles
- **Slider**: (e.g. https://vuetifyjs.com/en/components/sliders/#icons)
- **Slider:** the user can move the slider from "-" to "+"
- **Slider:** the slider in-/decreases the size of the tiles
- **Slider:** the slider provides 6 steps from the smallest to the biggest view
- **Tiles** consist of 2 parts
  - **Preview:** Contains the file-preview or icon and the checkbox
  - **Footer:** Contains the filetype-icon, file/folder name (evtl. truncated), 3-dots menu. **attention**: the 3-dots menu in the tiles-view contains the option "select" which triggers the checkbox to be a11y-compliant. update: other than in the design, we decided to NOT have the line with modified date within a tile.
- **Tiles**: There are 3 preview types
  1. **Images**: jpg, png, gif etc. image fits into the preview-container as `object-fit: cover`
  2. **Document-preview**: if supported, show a preview of the document. document-preview fits into the preview-container as `object-fit: scale-down`
  3. **Icon:** it not-previewable filetype: show only the Icon
- **Tiles**-interaction: click on preview performs the standard-action; click on checkbox checks, 3-dots opens the context menu.
- **Slider** Tiles-Sizes: the following width refers to the width of a single tile (width = 12 rem; width += 12rem):
  - Ratio: 10:9 (eg. 192 x 172,5 px)
  - Step 1: width: 12 rem (default)
  - Step 2: width: 24 rem
  - Step 3: width: 36 rem
  - Step 4: width: 48 rem
  - Step 5: width: 60 rem
  - Step 6: width: 72 rem
  - Steps are not available, if they dont fit into the content area.
- **Sorting:** Tiles can be sorted by a click on "Sort by: A-Z". Click opens a drop-menu with these options:
  - Name: "A-Z"
  - Name: "Z-A"
  - Modified: "Newest"
  - Modified: "Oldest"
  - Size: "Largest"
  - Size: "Smallest"
- **Pagination:** the tiles should not leave "blank" space at the end of the page
(https://en.wikipedia.org/wiki/Widows_and_orphans)
  - idea: set the default pagination page size to 1.000

## Definition of done
- Functional requirements
[ ] functionality described in the user story works
[ ] acceptance criteria are fulfilled
- Quality
[ ] code review happened
[ ] CI is green
[ ] critical code received unit tests by the developer
[ ] automated tests passed (if automated tests are not available, this test needs to be created and passed
- Non-functional requirements
[ ] no sonar cloud issues

### Small (Default)
Grid view - small

### Medium
Grid view - medium

### Large
Grid view - Large

### Small: Selected
image

### Selected / Hover

**static states:**
image

### Sorting Dropmenu:
Grid view - sorting

## Mobile: switch list <-> tile view

https://user-images.githubusercontent.com/26610733/146025353-8b06ac79-2027-442e-8994-44c5ea35e2ea.mp4

## Related issues
https://github.com/owncloud/web/issues/2585
https://github.com/owncloud/web/issues/3341

### LINKED ISSUE #6378: Display simple files preview as grid
# Description

## User Stories

* > As an early adopter I'd like to see file previews in a grid view, so that I dont need to read filenames to find the right file.

## Acceptance Criteria

1. **mimetype Images/video**: jpg, png, gif etc. image fits into the preview-container as `object-fit: cover`
  2. **preview Documents**: if supported, show a preview of the document. document-preview fits into the preview-container as `object-fit: scale-down`
  3. **Show Icon:** if not-previewable filetype: show only the Icon
- Arrange previews in grid view (not in list)
- **Tile-Size**: the following width refers to the width of a single tile:
  - Ratio: 10:9 (eg. 192 x 172,5 px)
  - default width: 12 rem
  - hint: later we'll have a configurable width (slider)
- for now we'll accept to have "missing" tiles in the last row of a paginated page
- show the fielname
- click on tile or filename triggers the default file action

**Note - not included:**
❌ selection
❌ hover
❌ context menu
❌ lazyloading for tiles

## Definition of done
- Functional requirements
[ ] functionality described in the user story works
[ ] acceptance criteria are fulfilled
- Quality
[ ] codre review happened
[ ] CI is green
[ ] critical code received unit tests by the developer
[ ] automated tests passed (if automated tests are not available, this test needs to be created and passed
- Non-functional requirements
[ ] no sonar cloud issues

### PR DESCRIPTION
**Pull request by pascalwengerter**
## Related Issue
- Fixes #6378

## Screenshots (if appropriate):

Screenshot 2023-01-10 at 15 57 54

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

File: packages/web-app-files/src/components/FilesList/ResourceTable.vue
Comment (pascalwengerter) [inline packages/web-app-files/src/components/FilesList/ResourceTable.vue:3]: @kulmann @dschmidt @JanAckermann @JammingBen @lookacat looking for feedback here 🤓 

This is a primitive&naive first draft for switching the tiles view. The original idea was aiming for API-parity between the `<oc-table>` and a new `<oc-tiles>` component since a lot of potential shared logic (sorting, filtering, selectedFiles, context menu, sharing indicators, ...) live in the `<ResourceTable>` and would otherwise need duplication or abstraction. Do y'all see a smarter way? Because right now it feels like messy entanglement and I'm not sure if clean boundaries are possible^^

File: packages/design-system/src/components/OcTiles/OcTiles.vue
Comment (kulmann) [inline packages/design-system/src/components/OcTiles/OcTiles.vue:384]:
```suggestion
  justify-content: space-around;
```
is much more pleasant for the eye, but any time the last row is incomplete (i.e. not the same number of tiles as in the other rows) the remaining tiles are centered, which doesn't look good... maybe we need some magic dummy tiles.

**Review by kulmann — CHANGES_REQUESTED** [review]
Could you re-render with the updated tile size during modification of the slider value? The URL query param changes immediately when you reach the next tick on the slider, but the tiles only update after you finished sliding the slider.

The main concern I have with your proposal of calculating the tile sizes based on the viewport size is that it stops to work good as soon as the viewport width becomes VERY large. E.g. @dschmidt with his ultra wide screen would have quite large tiles in the smallest scale. That's the reason why we wanted to have tile sizes based on multiples of a certain REM value, not based on the viewport size. I still think that it's a good idea...

**Comment by dschmidt** [comment]
with rem option  in `ViewOptions.vue`:
Screenshot_20221123_101841

with percent option:
Screenshot_20221123_101449


I agree with @kulmann that the latter is waaay to big for the smallest setting

File: packages/design-system/src/components/OcTiles/OcTiles.vue
Comment (dschmidt) [reply packages/design-system/src/components/OcTiles/OcTiles.vue:384]: The horizontal spacing is nice, but then the vertical spacing doesn't match - with this change alone

**Comment by tbsbdr** [comment]
Refined proposal I discussed with @kulmann
- **Spacing:** use "space-around" to achieve same effect as in windows explorer or MacOS finder
- ***Spacing** vertical spacing between  resources should be the same as horizontal

https://user-images.githubusercontent.com/26610733/203627733-6b2fa857-ea44-480e-90b0-425d8ba1a8f5.mp4

File: packages/web-app-files/src/components/FilesList/ResourceTable.vue
Comment (lookacat) [reply packages/web-app-files/src/components/FilesList/ResourceTable.vue:3]: @pascalwengerter for me it sounds good this way, or what makes you think it is messy this way? Not sure if there are better ideas

Comment (lookacat) [reply packages/web-app-files/src/components/FilesList/ResourceTable.vue:3]: At least i didn't have to think for long to understand how this is supposed to work ^^

Comment (pascalwengerter) [reply packages/web-app-files/src/components/FilesList/ResourceTable.vue:3]: Just talked to @kulmann today, TL;DR is that this is the way to go (the other option would be adding lots of duplication which then would have to be abstracted away _again_)

File: packages/web-app-files/src/composables/filesList/index.ts
Comment (dschmidt) [inline packages/web-app-files/src/composables/filesList/index.ts:92]: Just a nitpick, we usually just re-rexport from `index.ts` and put composables in their own files.

File: packages/web-app-files/src/views/spaces/GenericSpace.vue
Comment (dschmidt) [inline packages/web-app-files/src/views/spaces/GenericSpace.vue:119]: Is the plan to use `component :is='...'` in the long run?

Comment (pascalwengerter) [reply packages/web-app-files/src/views/spaces/GenericSpace.vue:120]: Not 100% sure yet tbh, I faced some problems of the `GenericSpace` not handling events being thrown by the switched-out component. Up for discussion I suppose, I can see another possible strategy that's pretty lean code-wise evolve :)

File: packages/web-app-files/src/composables/filesList/index.ts
Comment (pascalwengerter) [reply packages/web-app-files/src/composables/filesList/index.ts:92]: Good point thx, I will split it up (and also add tests etc) when it's time ;P

**Comment by pascalwengerter** [comment]
@kulmann could I get a cursory review here (mostly about UI/UX aspects, the code changes will receive some what-goes-where refactoring and are not yet final)?
Not sure what is in scope of the linked ticket

Known issues as of now:
- [ ] Navigation in none-root space has some hickups, seems like the fileId gets added to the route at some point and brakes the router?

**Comment by pascalwengerter** [comment]
> @kulmann could I get a cursory review here (mostly about UI/UX aspects, the code changes will receive some what-goes-where refactoring and are not yet final)? Not sure what is in scope of the linked ticket
> 
> Known issues as of now:
> 
>     * [ ]  Navigation in none-root space has some hickups, seems like the fileId gets added to the route at some point and brakes the router?

Update - I investigated some more and navigation inside folder in a space _does_ work, but is horribly inefficient (browser freezes, and navigation takes place ~40s after clicking). My suspectis the dynamic `ref` implementation which is needed for the Contextmenu logic

**Comment by pascalwengerter** [comment]
> > @kulmann could I get a cursory review here (mostly about UI/UX aspects, the code changes will receive some what-goes-where refactoring and are not yet final)? Not sure what is in scope of the linked ticket
> > Known issues as of now:
> > ```
> > * [ ]  Navigation in none-root space has some hickups, seems like the fileId gets added to the route at some point and brakes the router?
> > ```
> 
> Update - I investigated some more and navigation inside folder in a space _does_ work, but is horribly inefficient (browser freezes, and navigation takes place ~40s after clicking). My suspectis the dynamic `ref` implementation which is needed for the Contextmenu logic

Update - starting it with `pnpm build:w` (production build) as pointed out by @dschmidt does the trick, apparently all the deprecation warnings just wrangle the thing down...

File: packages/design-system/src/components/OcTile/OcTile.vue
Comment (kulmann) [inline packages/design-system/src/components/OcTile/OcTile.vue:99]: Please keep the box-shadow as `none` like before, but add `outline: 1px solid var(--oc-color-border);` instead. Discussed it with @tbsbdr that's how we want to start. Already had too much harsh feedback with shadows in other places of oC Web. 😅

File: packages/web-app-files/src/components/AppBar/AppBar.vue
Comment (kulmann) [inline packages/web-app-files/src/components/AppBar/AppBar.vue:148]: Please change order to
1. condensed
2. default
3. tiles

File: packages/design-system/src/components/OcTile/OcTile.vue
Comment (kulmann) [inline packages/design-system/src/components/OcTile/OcTile.vue:39]:
```suggestion
          <oc-resource :resource="resource" :folder-link="resourceRoute" :is-thumbnail-displayed="false" @click="$emit('click')" />
```

Thumbnails must be disabled here. Otherwise you might have glitchy situations where the OcTile has a large preview at the top and a small thumbnail in the bottom area. Managed to create that situation, don't have a formula to reproduce it, sorry. 👀

**Review by kulmann — CHANGES_REQUESTED** [review]
Awesome state / progress! Some tiny nitpicks, see comments. And I have two issues that you might decide to solve in a followup (also see screenshot):
- text truncation happens too early, there is a lot of horizontal white space left.
- preview size is sometimes shrinked / stretched. That's weird. object-fit and aspect-ratio don't seem to have an effect, I can disable the props and nothing changes (I wanted to try out if `contain` looks better, but didn't happen).

Screenshot 2023-01-13 at 16 11 35

**Comment by tbsbdr** [comment]
Played around with gap- and tiles sizes. The easiest way to avoid to much unbalanced gaps is to **reduce the tiles size**.

## Tilesize 250px -> big gaps  ❌

https://user-images.githubusercontent.com/26610733/212416491-b8f96d88-45c9-4654-ad18-009554fe566c.mp4


## Tilesize 160px -> small gaps ✅ 

https://user-images.githubusercontent.com/26610733/212416345-57712d66-a027-408e-badc-aa66c9598a8a.mp4

```css
.oc-tiles {
  display: grid;
  grid-template-columns: repeat(auto-fill, 160px);
  row-gap: 32px;
  column-gap: 8px;
  justify-content: space-between;

  .oc-tiles-item {
    width: 160px;

    @media (max-width: $oc-breakpoint-xsmall-max) {
      width: 100%;
    }
  }
}
```

I also **removed the leading file-icon** from the small tiles. @kulmann @pascalwengerter I'm in favour of small tiles as long as we don't have the slider. Do you agree?

**Comment by ownclouders** [comment]
Results for <strong>acceptance oC10</strong> https://drone.owncloud.com/owncloud/web/31680/30/1
<details><summary>:boom: The acceptance tests failed on retry. Please find the screenshots inside ...</summary>

<p>


### webUISharingInternalUsersShareWithPage-shareWithUsers_feature-L78.png
  

### webUISharingInternalUsersToRootShareWithPage-shareWithUsers_feature-L68.png
  


</p></details>
:boom: The oc10SharingIntUsers3 tests pipeline failed. The build has been cancelled.

**Comment by tbsbdr** [comment]
@kulmann for discussion later:

1. **Icon:** Use `function-line` icon for tiles-view icon screenshot_000217
2. **column gap:** is not yet 8px screenshot_000218
3. **Space:** should they be aligned with tiles? 
screenshot_000219

**Comment by ownclouders** [comment]
Results for <strong>acceptance oCIS</strong> https://drone.owncloud.com/owncloud/web/31748/71/1
:boom: The oCISUploadMove tests pipeline failed. The build has been cancelled.

**Comment by ownclouders** [comment]
Results for <strong>acceptance oCIS</strong> https://drone.owncloud.com/owncloud/web/31748/70/1
:boom: The oCISSharingPublic3 tests pipeline failed. The build has been cancelled.

**Comment by ownclouders** [comment]
Results for <strong>acceptance oCIS</strong> https://drone.owncloud.com/owncloud/web/31748/69/1
:boom: The oCISSharingPublic2 tests pipeline failed. The build has been cancelled.

**Comment by ownclouders** [comment]
Results for <strong>acceptance oCIS</strong> https://drone.owncloud.com/owncloud/web/31748/72/1
:boom: The oCISTrashbinJourney tests pipeline failed. The build has been cancelled.

**Comment by tbsbdr** [comment]
@pascalwengerter as you said, setting the tiles-behaviour not via viewport but via container query could do the trick, especially for smaller screens < 1000px:
https://user-images.githubusercontent.com/26610733/213001285-df2ed354-c7ce-497d-8f50-2e81be244b21.mp4

**Comment by pascalwengerter** [comment]
> @pascalwengerter as you said, setting the tiles-behaviour not via viewport but via container query (i.e. based on fileslist-width) could do the trick, especially for smaller screens < 1000px:
> screenshot_000220.mp4

That indeed looks pretty dope! But you've used https://caniuse.com/?search=container%20query which will only land in Firefox in a bunch of weeks, right?

**Comment by tbsbdr** [comment]
correct, used https://caniuse.com/?search=container%20query
imo okay if we provide a not-so-nice looking fallback for "older" browsers

**Review by kulmann — APPROVED** [review]
Looks pretty good already. Needs a rebase and probably some vue3 adjustments. Please ping @dschmidt, @JammingBen or me if you want some help with that. Let's get this merged after the rebase and iterate further from there. Once this is merged I can create smaller scoped followup tickets.

File: packages/web-pkg/src/constants.ts
Comment (kulmann) [inline packages/web-pkg/src/constants.ts:7]: The oCIS thumbnail service creates a square image for this resolution and your css then stretches it to 16:9 aspect ratio. That looks off... as a workaround you could use a higher resolution for now, so that the thumbnail service doesn't crop to a square resolution. 1000x1000 should do. The thumbnail service only has some limited hardcoded solutions and picks the one that is closest to the one you requested.

**Review by kulmann — CHANGES_REQUESTED** [review]

File: packages/web-pkg/src/constants.ts
Comment (pascalwengerter) [reply packages/web-pkg/src/constants.ts:7]: Per https://github.com/owncloud/ocis/blob/b24d126b3086351200c382d2dd46dea5bb05a66b/services/thumbnails/pkg/config/defaults/defaultconfig.go#L40 this would mean 1920x1080 I guess? Quite some big files potentially, but we can go with that for the beginning, yes

**Comment by pascalwengerter** [comment]
Thanks @JammingBen for a second pair of eyes! @kulmann ready for final review and _please_ squash-merge this!

**Review by kulmann — APPROVED** [review]
😍

**Pull request merged**
